src/shelly_sw_service.c

```diff
@@ -100,6 +116,25 @@ static void shelly_sw_set_state_ctx(struct shelly_sw_service_ctx *ctx,
       mgos_set_timer(600, 0, do_reset, ctx);
     }
   }
+
+  if (cfg->auto_off && auto_off_timer_id >= 0)
+    reset_auto_off_timer();  // Cancel timer if state changes so that only the
+                             // last timer is triggered if state changes
+                             // multiple times
+
+  if (strcmp(source, "auto_off") == 0) return;
+
+  bool auto_off = (new_state ? cfg->auto_off : false);
+  int delay_seconds = auto_off_delay_seconds(cfg->auto_off_delay);
+  if (auto_off && delay_seconds > 0) {
+    auto_off_timer_id =
+        mgos_set_timer(delay_seconds * 1000, 0, do_auto_off, ctx);
+  }
+}
+
+static void do_auto_off(void *arg) {
+  struct shelly_sw_service_ctx *ctx = arg;
+  shelly_sw_set_state_ctx(ctx, false, "auto_off");
```

Comment:
check for cfg->auto_off again here, it's possible that mode is changed while the timer is running.

also set `s_auto_off_timer_id` to `MGOS_INVALID_TIMER_ID`, since timer id for one-shot timer is invalidated when it fires.